### DISCUSSION TITLE
Better dataformat for results and outcomes

### DISCUSSION THREAD

**Discussion by EwoutH** [quaquel/EMAworkbench#374] [Ideas]
Currently the results data is dumped into two dataframes:
```Python
experiments, outcomes = perform_experiments()
```
While not necessarily bad, you do lose a bit of information in the process:

- Metadata about the experiment performed (like the ranges of parameters, sampling methods, etc.)
- The difference between model uncertainties and levers (both are just columns in experiments)
- The direct pairing between the experiments and the outcomes

While you can work around everything, it sometimes feels a bit inconvenient. For examples:
- You want to only explore your uncertainties, before exploring your policies, you have to manually select what your uncertainties were
- You can't easily fully reproduce an experiment from an resulting run
- You often have to provide functions with two data objects: Functions like `create_pivot_plot`, `CART` and `Prim` for example require an x and y object.
- "Successfull" outcomes have to be calculated separately

Therefore, I think there should be a more convenient dataformat possible that designs around these limitations.

Moreover, we might find some construct like a "score" or a "success definitions" useful, which can be included in this data structure.

**Comment by quaquel** [comment]
In my view, the issue is not with the return of `perform_experiments`, but rather in `save_results`. If you do perform_experiments, you have the model instance, and so most of the information you describe. This information, however, is lost when you do `save_results`

> You want to only explore your uncertainties, before exploring your policies, you have to manually select what your uncertainties were

Not true, you would do `perform_experiments(scenarios=100)`, and then levers would not be sampled.

> You can't easily fully reproduce an experiment from a resulting run

Yes, this is a key issue. First, it inks to spec-7 as well, which should be supported by the workbench. Second, you want metadata regarding the sampling technique that was used.

> Moreover, we might find some construct like a "score" or a "success definitions" useful, which can be included in this data structure.

In my personal experience, you generally play around with this during the analysis phase, and you can easily do so with the current dict of numpy arrays for outcomes.

**Reply by EwoutH** [reply]
Thanks for getting back. I'm going to think a bit further on this, but I think having it in a class could be beneficial. You can then to `results.experiments`, `results.outcomes`, `results.levers`, `results.model`, `results.n_scenarios`, etc., and have everything, consistently, in one place. Having it in a class also offers a lot of opportunities for filtering and scoring.

Ideally you could directly update a column also, so you say something like `results["success"] = ...`

Alternatively you could define a `success_function`.

```Python
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Callable, Union
import pandas as pd
import numpy as np

@dataclass
class Results:
    """Container for EMA experiment results that maintains experiment context and provides
    convenient access to different aspects of the results.
    
    Attributes
    ----------
    experiments : pd.DataFrame
        The input parameters for each experiment (both uncertainties and levers)
    outcomes : Dict[str, np.ndarray]
        The raw outcome arrays from the model
    uncertainties : List[str]
        Names of uncertainty parameters
    levers : List[str]
        Names of policy lever parameters
    scoring_function : Optional[Callable]
        Function that takes the combined DataFrame and returns a numeric score
    success_function : Optional[Callable]
        Function that takes the combined DataFrame and returns a boolean success indicator
    """
    
    experiments: pd.DataFrame
    outcomes: Dict[str, np.ndarray]
    uncertainties: List[str]
    levers: List[str]
    scoring_function: Optional[Callable] = None
    success_function: Optional[Callable] = None
    _data: Optional[pd.DataFrame] = field(init=False, default=None)
    
    def __post_init__(self):
        """Create the combined DataFrame and calculate score/success if functions provided"""
        # Create combined DataFrame
        self._data = self.experiments.copy()
        for outcome_name, values in self.outcomes.items():
            self._data[outcome_name] = values
            
        # Calculate score if function provided
        if self.scoring_function is not None:
            self._data['score'] = self.scoring_function(self._data)
            
        # Calculate success if function provided
        if self.success_function is not None:
            self._data['success'] = self.success_function(self._data)
    
    @property
    def data(self) -> pd.DataFrame:
        """The complete DataFrame containing experiments, outcomes, scores, and success"""
        return self._data
    
    @property
    def score(self) -> Optional[pd.Series]:
        """The calculated scores, if a scoring function was provided"""
        return self._data.get('score')
    
    @property
    def success(self) -> Optional[pd.Series]:
        """The success indicators, if a success function was provided"""
        return self._data.get('success')
    
    def set_scoring_function(self, func: Callable) -> None:
        """Set or update the scoring function and recalculate scores"""
        self.scoring_function = func
        self._data['score'] = func(self._data)
    
    def set_success_function(self, func: Callable) -> None:
        """Set or update the success function and recalculate success"""
        self.success_function = func
        self._data['success'] = func(self._data)
    
    def filter(self, condition: Union[pd.Series, Callable]) -> 'Results':
        """Create new Results object with filtered data
        
        Parameters
        ----------
        condition : Union[pd.Series, Callable]
            Either a boolean mask or a function that takes self.data and returns a mask
        """
        if callable(condition):
            mask = condition(self._data)
        else:
            mask = condition
            
        filtered_exp = self.experiments[mask]
        filtered_outcomes = {
            name: values[mask] for name, values in self.outcomes.items()
        }
        
        return Results(
            filtered_exp,
            filtered_outcomes,
            self.uncertainties,
            self.levers,
            self.scoring_function,
            self.success_function
        )

    def successful_only(self) -> 'Results':
        """Return only the successful experiments"""
        if 'success' not in self._data:
            raise ValueError("No success function defined")
        return self.filter(self._data['success'])
    
    def top_scores(self, n: int = 10) -> 'Results':
        """Return top n experiments by score"""
        if 'score' not in self._data:
            raise ValueError("No scoring function defined")
        return self.filter(self._data['score'].nlargest(n).index)
```

**Reply by quaquel** [reply]
Curious to see what you come up with. I have been thinking about this on and off for over ten years. I agree that some kind of `ExperimentalResults` class makes sense, but I never settled on a clean API. Moreover, having just the simple tuple unpacking with the experiments and outcomes has worked well. 

The main drawbacks in my view of the current code are:
1. No information on the experimental design used, nor on the seed of the rng. This basically makes the results not reproducible out of the box.
2. No information on the uncertainties and levers, and their data type and ranges.
3. No information on the outcomes and their data type
4. Need to write custom code for the serialization of the results (i.e., `save_results`).

So any solution should resolve these 4 issues. However, given the proven convenience of `experiments, outcomes = perform_experiments`, it probably should also support this style of tuple unpacking.

**Reply by EwoutH** [reply]
Thanks for the additional requirements!

I think I have an even better approach in the works in which we just subclass a Pandas DataFrame, and add some additional functionality to it (just like a GeoDataFrame extends a DataFrame). Then you can do `results.uncertainties` or `results.metrics` and only get the columns of that class. While still being able to use the familiar DataFrame interface. The model is also referenced, and for most functions it would require only to input this single Results object. 

It can get an `experiments_outcomes_tuple` attribute for backwards compatiblity.

As for naming hierarchy, I was considering the following:

![image](https://github.com/user-attachments/assets/19d3caeb-ce0d-48f9-8c95-2a5799da8612)

**Reply by quaquel** [reply]
I am not sure that subclassing pandas.DataFrame is the best way forward, but curious to see what you can come up with.

The main challenge you will run into is that outcomes can be a mix of 1d, 2d and n-d numpy arrays (and in the future possibly even time series with a dedicated index via a dataframe instead of a numpy array. 

By design, a data frame is built around a single array that contains all data, but given the described heterogeneity, you cannot do that here. 

Moreover, the memory requirement explodes. A numpy array requires a single block of memory. You can easily require several Gb to store all experiments and their results. If this can be split over multiple blocks of memory, that is easier to allocate.

**Reply by EwoutH** [reply]
That’s a fair point, hadn’t considered that. Thanks.

**Reply by EwoutH** [reply]
This might actually be an opportunity to get some aggregation features built in. Outcomes could be stored in raw format, but also be added to an outcomes dataframe:
- single value results get added automatically
- nd arrays can be aggregated to a single value

This way you have one, convenient dataframe for basic usage, and full flexibility to handle the outcomes the way you want if needed.
